tree/dataframe/src/RSqliteDS.cxx

```diff
@@ -26,6 +32,328 @@
+   const char *zName,
+   sqlite3_file *pFile,
+   int flags,
+   int * /*pOutFlags*/)
+{
+   VfsRootFile *p = new (pFile) VfsRootFile();
+   p->pFile.pMethods = NULL;
+
+   static const sqlite3_io_methods io_methods = {
+      1,
+      VfsRdOnlyClose,
+      VfsRdOnlyRead,
+      VfsRdOnlyWrite,
+      VfsRdOnlyTruncate,
+      VfsRdOnlySync,
+      VfsRdOnlyFileSize,
+      VfsRdOnlyLock,
+      VfsRdOnlyUnlock,
+      VfsRdOnlyCheckReservedLock,
+      VfsRdOnlyFileControl,
+      VfsRdOnlySectorSize,
+      VfsRdOnlyDeviceCharacteristics,
+      NULL,
+      NULL,
+      NULL,
+      NULL,
+      NULL,
+      NULL
+   };
+
+   if (flags & SQLITE_OPEN_READWRITE)
+      return SQLITE_IOERR;
+   if (flags & SQLITE_OPEN_DELETEONCLOSE)
+      return SQLITE_IOERR;
+   if (flags & SQLITE_OPEN_EXCLUSIVE)
+      return SQLITE_IOERR;
+
+   Davix::DavixError *err = NULL;
+   p->fd = p->pos.open(NULL, zName, O_RDONLY, &err);
+
+   if (!p->fd) {
+      printf("%s\n", err->getErrMsg().c_str());
+      return SQLITE_IOERR;
+   }
+
+   struct stat buf;
+   if (p->pos.stat(NULL, zName, &buf, NULL) == -1) {
+      return SQLITE_IOERR;
+   }
+   p->size = buf.st_size;
+
+   p->pFile.pMethods = &io_methods;
+   return 0;
+}
+
+static int VfsRdOnlyDelete(
+   sqlite3_vfs* __attribute__((unused)),
+   const char * /*zName*/,
+   int /*syncDir*/)
+{
+   return SQLITE_IOERR_DELETE;
+}
+
+static int VfsRdOnlyAccess(
+   sqlite3_vfs * /*vfs*/,
+   const char * /*zPath*/,
+   int flags,
+   int *pResOut)
+   {
+      *pResOut = 0;
+      if (flags == SQLITE_ACCESS_READWRITE) {
+         return SQLITE_OPEN_READONLY;
+      }
+   return SQLITE_OK;
+   }
+
+int VfsRdOnlyFullPathname(
+   sqlite3_vfs * /*vfs*/,
+   const char *zPath,
+   int nOut,
+   char *zOut)
+{
+   zOut[nOut-1] = '\0';
+   sqlite3_snprintf(nOut, zOut, "%s", zPath);
+   return SQLITE_OK;
+}
+
+/**
+ * Taken from unixRandomness
+ */
+static int VfsRdOnlyRandomness(
+   sqlite3_vfs * /*vfs*/,
+   int nBuf,
+   char *zBuf)
+{
+   assert((size_t)nBuf >= (sizeof(time_t) + sizeof(int)));
+   memset(zBuf, 0, nBuf);
+   pid_t randomnessPid = getpid();
+   int fd, got;
+   fd = open("/dev/urandom", O_RDONLY, 0);
+   if (fd < 0) {
+     time_t t;
+     time(&t);
+     memcpy(zBuf, &t, sizeof(t));
+     memcpy(&zBuf[sizeof(t)], &randomnessPid, sizeof(randomnessPid));
+     assert(sizeof(t) + sizeof(randomnessPid) <= (size_t)nBuf);
+     nBuf = sizeof(t) + sizeof(randomnessPid);
+   } else {
+     do {
+       got = read(fd, zBuf, nBuf);
+     } while (got < 0 && errno == EINTR);
+     close(fd);
+   }
+  return nBuf;
+}
+
+static int VfsRdOnlySleep(
+   sqlite3_vfs * /*vfs*/,
+   int microseconds)
+{
+   gSystem->Sleep(microseconds / 1000);
+   return microseconds;
+}
+
+static int VfsRdOnlyGetLastError(
+   sqlite3_vfs * /*vfs*/,
+   int /*not_used1*/,
+   char * /*not_used2*/)
+{
+   return 0;
+}
+
+/**
+ * Taken from unixCurrentTimeInt64()
+ */
+static int VfsRdOnlyCurrentTimeInt64(
```

Comment:
For the implementation of this callback, we should take a look into `std::chrono`. The current implementation is tied to UNIX.

Reply:
Which platforms are we losing that have Davix support but not `gettimeofday`?

Reply:
None, I think, but I thought it would be nice not to bind ourselves to UNIX otherwise. On the other hand, it seems that only C++20 provides the tools to properly replace `gettimeofday`, so we could add a comment for future work.

Reply:
I think we could get by with the two functions below, no?
https://en.cppreference.com/w/cpp/chrono/system_clock/to_time_t
https://en.cppreference.com/w/cpp/chrono/c/localtime